Pal/src/host/Linux-SGX/tools/ra-tls/secret_prov_attest.c

```diff
@@ -33,39 +33,58 @@
 #include "secret_prov.h"
 #include "util.h"
 
-/* these are globals because the user may continue using the SSL session even after invoking
- * secret_provision_start() (in the user-supplied callback) */
-static mbedtls_ctr_drbg_context g_ctr_drbg;
-static mbedtls_entropy_context g_entropy;
-static mbedtls_ssl_config g_conf;
-static mbedtls_x509_crt g_verifier_ca_chain;
-static mbedtls_net_context g_verifier_fd;
-static mbedtls_ssl_context g_ssl;
-static mbedtls_pk_context g_my_ratls_key;
-static mbedtls_x509_crt g_my_ratls_cert;
-
-static uint8_t* provisioned_secret = NULL;
-static size_t provisioned_secret_size = 0;
+static uint8_t* g_provisioned_secret = NULL;
+static size_t g_provisioned_secret_size = 0;
 
 int secret_provision_get(uint8_t** out_secret, size_t* out_secret_size) {
     if (!out_secret || !out_secret_size)
         return -EINVAL;
 
-    *out_secret      = provisioned_secret;
-    *out_secret_size = provisioned_secret_size;
+    *out_secret      = g_provisioned_secret;
+    *out_secret_size = g_provisioned_secret_size;
     return 0;
 }
 
 void secret_provision_destroy(void) {
-    if (provisioned_secret && provisioned_secret_size)
+    if (g_provisioned_secret && g_provisioned_secret_size)
 #ifdef __STDC_LIB_EXT1__
-        memset_s(provisioned_secret, 0, provisioned_secret_size);
+        memset_s(g_provisioned_secret, 0, g_provisioned_secret_size);
 #else
-        memset(provisioned_secret, 0, provisioned_secret_size);
+        memset(g_provisioned_secret, 0, g_provisioned_secret_size);
 #endif
-    free(provisioned_secret);
-    provisioned_secret      = NULL;
-    provisioned_secret_size = 0;
+    free(g_provisioned_secret);
+    g_provisioned_secret      = NULL;
+    g_provisioned_secret_size = 0;
+}
+
+extern int common_write(mbedtls_ssl_context* ssl, const uint8_t* buf, size_t size);
+extern int common_read(mbedtls_ssl_context* ssl, uint8_t* buf, size_t size);
+extern int common_close(mbedtls_ssl_context* ssl);
+
+int secret_provision_write(struct ra_tls_ctx* ctx, const uint8_t* buf, size_t size) {
+    mbedtls_ssl_context* ssl = (mbedtls_ssl_context*)ctx->ssl;
+    return common_write(ssl, buf, size);
+}
+
+int secret_provision_read(struct ra_tls_ctx* ctx, uint8_t* buf, size_t size) {
+    mbedtls_ssl_context* ssl = (mbedtls_ssl_context*)ctx->ssl;
+    return common_read(ssl, buf, size);
+}
+
+int secret_provision_close(struct ra_tls_ctx* ctx) {
+    mbedtls_ssl_context* ssl = (mbedtls_ssl_context*)ctx->ssl;
+    mbedtls_ssl_config* conf = (mbedtls_ssl_config*)ctx->conf;
+    mbedtls_net_context* net = (mbedtls_net_context*)ctx->net;
+
+    int ret = common_close(ssl);
+
+    mbedtls_ssl_free(ssl);
+    mbedtls_ssl_config_free(conf);
+    mbedtls_net_free(net);
+    free(ssl);
+    free(conf);
+    free(net);
+    return ret;
```

Comment:
consider to zero the variables before return.
ctx->net = 0;
ctx->ssl = 0;
...